flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -284,6 +284,97 @@ arrayLoadExtValue(fir::FirOpBuilder &builder, mlir::Location loc,
   TODO(loc, "component is boxed, retreive its type parameters");
 }
 
+/// Place \p exv in memory if it is not already a memory reference. If
+/// \p forceValueType is provided, the value is first casted to the provided
+/// type before being stored (this is mainly intended for logicals whose value
+/// may be `i1` but needed to be stored as Fortran logicals).
+static fir::ExtendedValue
+placeScalarValueInMemory(fir::FirOpBuilder &builder, mlir::Location loc,
+                         const fir::ExtendedValue &exv,
+                         llvm::Optional<mlir::Type> forceValueType) {
+  mlir::Value valBase = fir::getBase(exv);
+  // Functions are always referent.
+  if (valBase.getType().template isa<mlir::FunctionType>() ||
+      fir::conformsWithPassByRef(valBase.getType()))
+    return exv;
+
+  assert(!fir::hasDynamicSize(valBase.getType()) &&
+         "only expect lengthless scalars to be by value");
+
+  // Since `a` is not itself a valid referent, determine its value and
+  // create a temporary location at the begining of the function for
+  // referencing.
+  mlir::Value val =
+      forceValueType
+          ? builder.createConvert(loc, forceValueType.getValue(), valBase)
+          : valBase;
+  mlir::FuncOp func = builder.getFunction();
+  auto initPos = builder.saveInsertionPoint();
+  builder.setInsertionPointToStart(&func.front());
```

Comment:
There should already be code somewhere to add these temporaries for value to memory reference conversions on call interfaces.

By the way, they require a special flag so we know they have a limited lifetime.

Reply:
This is moved code so that it could be shared, although I agree it should use `builder.createTemporary()`. I have also added the `AdaptToByRefAttr` flag.